# XML, HTML, and Web Scraping

JSON and XML are two different ways to represent hierarchical data. Which one is better? There are lots of articles online which discuss similarities and differences between JSON and XML and their advantages and disadvantages. Both formats are still in current usage, so it is good to be familiar with both. However, JSON is more common, so we'll focus on working with JSON representations of hierarchical data.

The reading covered an example of using Beautiful Soup to parse XML. Rather than doing another example XML now, we'll skip straight to scraping HTML from a webpage. Both HTML and XML can be parsed in a similar way with Beautiful Soup.

In [11]:
import pandas as pd

## Scraping an HTML table with Beautiful Soup

Open the URL https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population and scroll down until you see a table of the cities in the U.S. with population over 100,000 (as of Jul 1, 2022). We'll use Beautiful Soup to scrape information from this table.

Read in the HTML from the ULR using the `requests` library.

In [12]:
import requests
response = requests.get('https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population')

Use Beautiful Soup to parse this string into a tree called `soup`

In [13]:
from bs4 import BeautifulSoup
soup = BeautifulSoup(response.content, "html.parser")

To find an HTML tag corresponding to a specific element on a webpage, right-click on it and choose "Inspect element". Go to the cities table Wikipedia page and do this now.

You should find that the cities table on the Wikipedia page corresponds to the element

```
<table class="wikitable sortable jquery-tablesorter" style="text-align:center">
```

There are many `<table>` tags on the page.

In [5]:
len(soup.find_all("table"))

10

We can use attributes like `class=` and `style=` to narrow down the list.

In [15]:
len(soup.find_all("table",
                  attrs={
                      "class": "wikitable",
                      "style": "text-align:right"}
                  ))

3

At this point, you can manually inspect the tables on the webpage to find that the one we want is the first one (see `[0]` below). We'll store this as `table`.

In [16]:
table = soup.find_all("table",
                  attrs={
                      "class": "wikitable",
                      "style": "text-align:right"}
                  )[0]

**Now you will write code to scrape the information in `table` to create a Pandas data frame with one row for each city and columns for: city, state, population (2022 estimate), and 2020 land area (sq mi).** Refer to the Notes/suggestions below as you write your code. A few Hints are provided further down, but try coding first before looking at the hints.

Notes/suggestions:

- Use as a guide the code from the reading that produced the data frame of Statistics faculty
- Inspect the page source as you write your code
- You will need to write a loop to get the information for all cities, but you might want to try just scraping the info for New York first
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `.get_text(strip = True)` instead of `.text`
- Don't forget to convert to a Pandas Data Frame; it should have 333 rows and 4 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it --- e.g., what is the population density for all cities in CA? --- then you would need to clean the data first (to clean strings and convert to quantitative). (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)

In [ ]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED

In [19]:
# initialize an empty list
rows = []

for row in table.find_all("tr")[1:]:
    cells = row.find_all('td')

    if len(cells) >= 7:
        city_name = cells[0].find('a').get_text(strip=True)   # City name (inside <a> tag)
        state = cells[1].get_text(strip=True)                 # State name
        population = cells[3].get_text(strip=True)            # Population (2020 estimate)
        land_area = cells[5].get_text(strip=True)             # Land area (2020)(sq mi)

        # Append as a dictionary to the rows list
        rows.append({
            "City": city_name,
            "State": state,
            "Population": population,
            "Land Area (sq mi)": land_area
        })

#Convert into a Pandas DataFrame
df = pd.DataFrame(rows)

print(df)

            City State Population Land Area (sq mi)
0       New York    NY  8,804,190             300.5
1    Los Angeles    CA  3,898,747             469.5
2        Chicago    IL  2,746,388             227.7
3        Houston    TX  2,304,580             640.4
4        Phoenix    AZ  1,608,139             518.0
..           ...   ...        ...               ...
331         Yuma    AZ     95,548             120.7
332  New Bedford    MA    101,079              20.0
333      Suffolk    VA     94,324             399.2
334     Hesperia    CA     99,818              72.7
335    Davenport    IA    101,724              63.8

[336 rows x 4 columns]


Hints:

- Each city is a row in the table; find all the `<tr>` tags to find all the cities
- Look for the `<td>` tag to see table entries within a row
- The rank column is represented by `<th>` tags, rather than `<td>` tags. So within a row, the first (that is, `[0]`) `<td>` tag corresponds to the city name.

## Aside: Scraping an HTML table with Pandas



The Pandas command `read_html` can be used to scrape information from an HTML table on a webpage.

We can call `read_html` on the URL.

In [43]:
#pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

However, this scrapes all the tables on the webpage, not just the one we want. As with Beautiful Soup, we can narrow the search by specifying the table attributes.

In [44]:
#pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population", attrs = {'class': 'wikitable', "style": "text-align:right"})

This still returns 3 tables. As we remarked above, the table that we want is the first one (see `[0]` below).

In [45]:
df_cities2 = pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population", attrs = {'class': 'wikitable', "style": "text-align:right"})[0]
#df_cities2

Wait, that seemed much easier than using Beautiful Soup, and it returned a data frame, and we even got for free some formatting like removing the commas from the population! Why didn't we just use `read_html` in the first place? It's true the `read_html` works well when scraping information from an HTML *table*. Unfortunately, you often want to scrape information from a webpage that isn't conveniently stored in an HTML table, in which case `read_html` won't work. (It only searches for `<table>`, `<th>`, `<tr>`, and `<td>` tags, but there are many other HTML tags.) Though Beautiful Soup is not as simple as `read_html`, it is more flexible and thus more widely applicable.

## Scraping information that is NOT in a `<table>` with Beautiful Soup

The Cal Poly course catalog http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory contains a list of courses offered by the Statistics department. **You will scrape this website to obtain a Pandas data frame with one row for each DATA or STAT course and two columns: course name and number (e.g, DATA 301. Introduction to Data Science) and term typically offered (e.g., Term Typically Offered: F, W, SP).**

Note: Pandas `read_html` is not help here since the courses are not stored in a `<table>.`

In [46]:
#pd.read_html("http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")


Notes/suggestions:


- Inspect the page source as you write your code
- The courses are not stored in a `<table>`. How are they stored?
- You will need to write a loop to get the information for all courses, but you might want to try just scraping the info for DATA 100 first
- What kind of tag is the course name stored in? What is the `class` of the tag?
- What kind of tag is the quarter(s) the course is offered stored in? What is the `class` of the tag? Is this the only tag of this type with the class? How will you get the one you want?
- You don't have to remove the number of units (e.g., 4 units) from the course name and number, but you can try it if you want
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `get_text(strip = True)` instead of `text`
- Don't forget to convert to a Pandas Data Frame; it should have 74 rows and 2 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it then you might need to clean the data first. (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)



In [ ]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED

In [40]:
url = 'http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory'
response = requests.get(url)

soup = BeautifulSoup(response.content, 'html.parser')

course_blocks = soup.find_all("div", {"class": "courseblock"})

courses = []

for course_block in course_blocks:
    title_block = course_block.find("p", {"class": "courseblocktitle"})
    course_title = title_block.get_text(strip=True)

    if course_title.startswith("DATA") or course_title.startswith("STAT"):
        parts = course_title.split(" ", 2)
        course_number_and_name = f"{parts[0]} {parts[1]}. {parts[2]}"

        course_number_and_name = course_title.rsplit('.', 1)[0].strip()  # Keep everything before the last dot

        term_block = course_block.find("p", {"class": "noindent"})
        term_offered = term_block.get_text(strip=True).replace("Term Typically Offered: ", "") if term_block else "N/A"


        courses.append({
            "Course Number and Name": course_number_and_name.strip(),  # Clean any extra whitespace
            "Term Typically Offered": term_offered
        })

df = pd.DataFrame(courses)

print(df)


###Chat GPT was used to split the units from the name and number properly
###Chat GPT was used to split DATA and STAT and the parts within the titles

                               Course Number and Name Term Typically Offered
0                    DATA 100. Data Science for All I               F, W, SP
1              DATA 301. Introduction to Data Science               F, W, SP
2           DATA 401. Data Science Process and Ethics                      F
3   DATA 402. Mathematical Foundations of Data Sci...                      F
4          DATA 403. Data Science Projects Laboratory                      F
..                                                ...                    ...
69              STAT 551. Statistical Learning with R                      F
70            STAT 566. Graduate Consulting Practicum                     SP
71                 STAT 570. Selected Advanced Topics                    TBD
72           STAT 590. Graduate Seminar in Statistics               F, W, SP
73                                   STAT 599. Thesis                    TBD

[74 rows x 2 columns]


Hints:

- Each course is represented by a `<div>` with `class=courseblock`, so you can find all the courses with `soup.find_all("div", {"class": "courseblock"})`
- The course name is in a `<p>` tag with `class=courseblocktitle`, inside a `<strong>` tag. (Though I don't think we need to find the strong tag here.)
- The term typically offered is in `<p>` tag with `class=noindent`. However, there are several tags with this class; term typically offered is the first one.
- If you want to use Beautiful Soup to remove the course units (e.g., 4 units), find the `<span>` tag within the course name tag and `.extract()` this span tag